### PR TITLE
Sanitize Problems service API surface

### PR DESCRIPTION
**Pull request by donat**
Fixes https://github.com/gradle/gradle-private/issues/3947

The API surface of the `Problems` interface is too big and there are still multiple ways to do one thing. This PR fixes that with the following changes:
- Move internal methods to `InternalProblems`
- Move `report()` method from builder to a new `Problem` subtype called `ReportableProblem`
- replace throwIt() with ProblemBuilder.throwing()

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test QFL

**Comment by donat** [comment]
@bot-gradle test QFL

File: subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java [hunk 1]
```diff
@@ -231,7 +229,7 @@ public void report() {
     }
 
     public RuntimeException throwIt() {
-        throw throwError(exception, buildInternal(ERROR));
+        throw throwError(exception, buildInternal(ERROR, false));
     }
```
Comment (hegyibalint) [inline subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java:233]: 🤔 Maybe this should be removed, as the action-based throwing is in place?

File: subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java [hunk 2]
```diff
@@ -244,6 +242,10 @@ public RuntimeException throwError(RuntimeException exception, Problem problem)
...
 
     private void report(Problem problem) {
-        buildOperationProgressEventEmitter.emitNowIfCurrent(problem);
+        problemsService.reportAsProgressEvent(problem);
+    }
```
Comment (hegyibalint) [inline subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java:246]: 🤔 Maybe this should be removed, and we should rely on `ReportableProblem` to make the reporting.

**Review by hegyibalint — APPROVED** [review]

File: subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java (hunk 1 above)
Comment (donat) [reply subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java:233]: Actually, this is unused code, thanks for pointing this out.

File: subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java (hunk 2 above)
Comment (donat) [reply subprojects/problems/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java:246]: With the correct approach there are reportable and the others are not. You can create `ReportableProbem` with Problems.createProblemBuilder(), but not within `throwing()` nor in `rethrowing()`

**Comment by donat** [comment]
@bot-gradle test and merge

**Pull request merged**